File: lib/lux/prism.ex
defmodule Lux.Prism do
  @moduledoc """
  Modular, composable units of functionality for defining actions.

  Prisms are used to define actions that can be executed by agents.

  ## Example

      defmodule MyApp.Prisms.WeatherPrism do
        use Lux.Prism,
          name: "Weather Data",
          description: "Fetches weather data for a given location",
          input_schema: %{
            type: :object,
            properties: %{
              location: %{type: :string, description: "City name"},
              units: %{type: :string, description: "Temperature units (C/F)"}
            }
          }

        def handler(input, _ctx) do
          # Implementation
        end
      end

  """
  use Lux.Types

  @typedoc """
  A schema is either a map of key-value pairs that describe the structure of the data,
  or a module that implements the Lux.SignalSchema behaviour.
  """
  @type schema :: map() | module()

  @typedoc """
  A handler is a function or a module that handles the data.
  """
  @type handler :: function() | mfa() | binary()

  @typedoc """
  A validator is a function or a module that validates the data.
  """
  @type validator :: function() | mfa() | binary()

  defstruct [
    :id,
    :name,
    :handler,
    :description,
    :examples,
    :input_schema,
    :output_schema
  ]

  @type t :: %__MODULE__{
          id: String.t(),
          name: String.t(),
          handler: handler(),
          description: nullable(String.t()),
          examples: nullable([String.t()]),
          input_schema: nullable(schema()),
          output_schema: nullable(schema())
        }

  @doc """
  Creates a new prism from a map or keyword list.
  """
  @spec new(map() | keyword()) :: t()
  def new(attrs) when is_map(attrs) do
    %__MODULE__{
      id: attrs[:id] || "",
      name: attrs[:name] || "",
      handler: attrs[:handler] || nil,
      description: attrs[:description] || "",
      examples: attrs[:examples] || [],
      input_schema: resolve_schema(attrs[:input_schema]),
      output_schema: resolve_schema(attrs[:output_schema])
    }
  end

  def new(attrs) when is_list(attrs) do
    attrs |> Map.new() |> new()
  end

  @doc """
  A handler is the function that will be called when the prism is executed.
  """
  @callback handler(input :: any(), context :: any()) :: {:ok, any()} | {:error, any()}

  defmacro __using__(opts) do
    quote do
      @behaviour Lux.Prism

      alias Lux.Prism

      # Register compile-time attributes
      Module.register_attribute(__MODULE__, :prism_config, persist: false)
      Module.register_attribute(__MODULE__, :prism_struct, persist: false)

      # Store the configuration at compile time
      @prism_config %{
        name: Keyword.get(unquote(opts), :name, __MODULE__ |> Module.split() |> List.last()),
        description: Keyword.get(unquote(opts), :description, ""),
        input_schema: Keyword.get(unquote(opts), :input_schema),
        output_schema: Keyword.get(unquote(opts), :output_schema),
        examples: Keyword.get(unquote(opts), :examples, []),
        id: Keyword.get(unquote(opts), :id, Lux.UUID.generate())
      }

      # Create the struct at compile time
      @prism_struct Lux.Prism.new(
                      id: Lux.UUID.generate(),
                      name: @prism_config.name,
                      description: @prism_config.description,
                      input_schema: @prism_config.input_schema,
                      output_schema: @prism_config.output_schema,
                      examples: @prism_config.examples
                    )

      def run(input, context \\ nil) do
        Lux.Prism.run(__MODULE__, input, context)
      end

      @doc """
      Returns the Prism struct for this module.
      """
      def view do
        %{@prism_struct | handler: &__MODULE__.handler/2}
      end

      defoverridable run: 2
    end
  end

  def run(schema, input, context \\ nil)

  def run(module, input, context) when is_atom(module) do
    module.handler(input, context)
  end

  def run(%__MODULE__{handler: handler}, input, context) do
    handler.(input, context)
  end

  @optional_callbacks []

  @doc """
  Resolves a schema reference to its actual schema definition.
  If the schema is a module that implements Lux.SignalSchema, returns its schema.
  Otherwise, returns the schema as is.
  """
  def resolve_schema(schema) when is_atom(schema) do
    if function_exported?(schema, :schema, 0) do
      schema.schema()
    else
      schema
    end
  end

  def resolve_schema(schema), do: schema
end

File: lux/priv/python/lux/prism.py
from abc import ABC, abstractmethod
from erlport.erlterms import Atom
import uuid
import os
import re

class Prism(ABC):
    def __init__(self, **kwargs):
        if 'id' not in kwargs and hasattr(self, 'id'):
            kwargs['id'] = self.id
        if 'name' not in kwargs and hasattr(self, 'name'):
            kwargs['name'] = self.name
        if 'description' not in kwargs and hasattr(self, 'description'):
            kwargs['description'] = self.description
        if 'input_schema' not in kwargs and hasattr(self, 'input_schema'):
            kwargs['input_schema'] = self.input_schema
        if 'output_schema' not in kwargs and hasattr(self, 'output_schema'):
            kwargs['output_schema'] = self.output_schema

        if 'name' in kwargs:
            pattern = re.compile(r'^[A-Z][A-Za-z0-9]*(?:\.[A-Z][A-Za-z0-9]*)*$')
            if not pattern.match(kwargs['name']):
                raise ValueError("Name must follow namespace convention (e.g. 'Lux.Prisms.MyPrism')")

        self.id = kwargs.get('id', str(uuid.uuid4()))
        self.name = kwargs.get('name', '')
        self.description = kwargs.get('description', '')
        self.input_schema = kwargs.get('input_schema', {})
        self.output_schema = kwargs.get('output_schema', {})
    
    def view(self):
        return {
            '__class__': self.__class__.__name__,
            '__struct__': Atom(b'Elixir.Lux.Prism'),
            'id': self.id,
            'name': self.name,
            'examples': '',
            'description': self.description,
            'input_schema': self.input_schema,
            'output_schema': self.output_schema,
            'handler': (Atom(b'python'), os.path.abspath(__file__))
        }

    @abstractmethod
    def handler(self, input, context):
        pass

    @staticmethod
    @abstractmethod
    def new():
        pass